File: src/lib/project_store.py
"""Project store for Zephyr Desktop.

Wraps ConfigManager to provide CRUD operations on projects,
persisted to projects.json in the config directory.
"""

from datetime import datetime, timezone

from src.lib.config_manager import ConfigManager
from src.lib.models import ProjectConfig

PROJECTS_FILE = "projects.json"


class ProjectStore:
    """CRUD store for ProjectConfig instances backed by projects.json.

    Args:
        config_manager: ConfigManager instance for file I/O.
    """

    def __init__(self, config_manager: ConfigManager):
        self._cm = config_manager

    def _load(self) -> dict[str, dict]:
        """Load the raw projects dict keyed by project id."""
        data = self._cm.load_json(PROJECTS_FILE)
        return data.get("projects", {})

    def _save(self, projects: dict[str, dict]) -> None:
        """Persist the projects dict."""
        self._cm.save_json(PROJECTS_FILE, {"projects": projects})

    def list_projects(self) -> list[ProjectConfig]:
        """Return all projects sorted by name."""
        projects = self._load()
        result = [ProjectConfig.from_dict(p) for p in projects.values()]
        result.sort(key=lambda p: p.name.lower())
        return result

    def get_project(self, project_id: str) -> ProjectConfig | None:
        """Return a single project by ID, or None if not found."""
        projects = self._load()
        data = projects.get(project_id)
        if data is None:
            return None
        return ProjectConfig.from_dict(data)

    def add_project(self, project: ProjectConfig) -> None:
        """Add a new project. Raises ValueError if ID already exists."""
        projects = self._load()
        if project.id in projects:
            raise ValueError(f"Project with id '{project.id}' already exists")
        projects[project.id] = project.to_dict()
        self._save(projects)

    def update_project(self, project: ProjectConfig) -> None:
        """Update an existing project. Raises KeyError if not found."""
        projects = self._load()
        if project.id not in projects:
            raise KeyError(f"Project with id '{project.id}' not found")
        project.updated_at = datetime.now(timezone.utc).isoformat()
        projects[project.id] = project.to_dict()
        self._save(projects)

    def remove_project(self, project_id: str) -> None:
        """Remove a project by ID. Raises KeyError if not found."""
        projects = self._load()
        if project_id not in projects:
            raise KeyError(f"Project with id '{project_id}' not found")
        del projects[project_id]
        self._save(projects)

File: src/services/project-store.ts
/**
 * ProjectStore — CRUD operations for persisted ProjectConfig records.
 *
 * Runs in the Electron main process. All reads and writes go through
 * ConfigManager so data is stored atomically in ~/.zephyr/projects.json.
 *
 * Why a separate service from ConfigManager: ConfigManager handles raw JSON
 * I/O; ProjectStore owns the domain logic — ID generation, timestamp updates,
 * duplicate detection, and not-found handling.
 */

import { ProjectConfig, createProjectConfig } from '../shared/models';
import { ConfigManager } from './config-manager';

const PROJECTS_FILE = 'projects.json';

export class ProjectStore {
  private readonly config: ConfigManager;

  constructor(config: ConfigManager) {
    this.config = config;
  }

  /**
   * Returns all stored projects. Returns an empty array when no file exists yet.
   */
  listProjects(): ProjectConfig[] {
    return this.load();
  }

  /**
   * Returns a single project by ID, or null if not found.
   */
  getProject(id: string): ProjectConfig | null {
    return this.load().find((p) => p.id === id) ?? null;
  }

  /**
   * Adds a new project to the store.
   *
   * Assigns a UUID and timestamps if not already present.
   * Throws if a project with the same ID already exists (duplicate detection).
   *
   * @param config - Partial or full project configuration
   * @returns The fully populated ProjectConfig that was saved
   */
  addProject(config: Partial<ProjectConfig>): ProjectConfig {
    const projects = this.load();
    const newProject = createProjectConfig(config);

    const duplicate = projects.find((p) => p.id === newProject.id);
    if (duplicate) {
      throw new Error(`Project with id "${newProject.id}" already exists`);
    }

    projects.push(newProject);
    this.save(projects);
    return newProject;
  }

  /**
   * Merges partial changes into an existing project and updates `updated_at`.
   *
   * Throws if the project is not found.
   *
   * @param id - The UUID of the project to update
   * @param partial - Fields to merge into the existing record
   * @returns The updated ProjectConfig
   */
  updateProject(id: string, partial: Partial<ProjectConfig>): ProjectConfig {
    const projects = this.load();
    const index = projects.findIndex((p) => p.id === id);

    if (index === -1) {
      throw new Error(`Project with id "${id}" not found`);
    }

    const updated: ProjectConfig = {
      ...projects[index],
      ...partial,
      // Preserve the original id — callers must not change it via update
      id,
      // Always refresh updated_at on every update
      updated_at: new Date().toISOString(),
    };

    projects[index] = updated;
    this.save(projects);
    return updated;
  }

  /**
   * Removes a project by ID.
   *
   * @returns true if the project was found and removed, false if not found
   */
  removeProject(id: string): boolean {
    const projects = this.load();
    const index = projects.findIndex((p) => p.id === id);

    if (index === -1) {
      return false;
    }

    projects.splice(index, 1);
    this.save(projects);
    return true;
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private load(): ProjectConfig[] {
    return this.config.loadJson<ProjectConfig[]>(PROJECTS_FILE) ?? [];
  }

  private save(projects: ProjectConfig[]): void {
    this.config.saveJson(PROJECTS_FILE, projects);
  }
}
